Create a GitHub action that will allow creating releases

ewanharris opened the pull request:
This PR introduces a GitHub action that will allow releases to be created solely from GitHub, the following was also done:

* To try and reduce duplication the existing PR build was split out into multiple [composite actions](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) for the Android, iOS, and packaging build steps.
* I removed the custom caching of node_modules in favour of the usage in the `setup-node` action
* I have also added some documentation related to the GitHub actions setup in the repo to cover the usage of these composite actions and describe how the existing CI setup functions.

Before the release workflow is usable the following needs to be done

- [x] Configure an environment called `Release`
- [x] Configure that environment to require an approval for deploy from the SDK team

You can see an example of this on my repo [here](https://github.com/ewanharris/titanium_mobile/actions/runs/1741344729)

ewanharris committed:
ci: split out android job into a composite action

ewanharris committed:
ci: move ios and package to composite actions

ewanharris committed:
ci: add release action

ewanharris committed:
ci: allow inputting branch and release type

ewanharris committed:
ci: hand pre-releases correctly

ewanharris committed:
docs: add some docs on the ci process

ewanharris committed:
ci: use newer setup-node action that caches global npm cache

Maybe faster?

ewanharris committed:
docs: add docs on releasing the sdk

build commented:
<!--
  0 failure: 
  2 warning:  Tests have failed..., There is no linke...
  3 messages
  1 markdown notices
  DangerID: danger-id-Danger;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>There is no linked JIRA ticket in the PR body. Please include the URL of the relevant JIRA ticket. If you need to, you may file a ticket on <a href="https://jira.appcelerator.org/secure/CreateIssue!default.jspa">JIRA</a></td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 1 tests have failed
There are 1 tests failing and 1157 skipped out of 20851 total tests.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-13252/2/artifact/dist/mobilesdk-10.2.1.v20220124101308-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.emulator.main.Titanium.Network.HTTPClient</td><td>responseXML (12)</td><td>10.22</td><td><pre>Error: failed to retrieve RSS feed: [object Object]
at HTTPClient.xhr.onerror (/ti.network.httpclient.test.js:42:8)</pre></td></tr>
</table>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against f5a3aa57f3465f4ef73394577099d8acde1a998f
</p>

hansemannn commented:
@ewanharris Let me know if there is something we can help with here!

ewanharris commented:
@hansemannn  I think @cb1kenobi  had this in his backlog to review, but if you also wanted to take a look through and ask any questions you might have about how it all works (maybe there's places in the docs I could clarify better?) that would be awesome.

cb1kenobi commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,165 @@
+    name: Package
+    env:
+      SDK_BUILD_CACHE_DIR: ${{ github.workspace }}/.native-modules
+      DEVELOPER_DIR: /Applications/Xcode_13.0.app/Contents/Developer
```
Shouldn't we be using Xcode 13.2.1?

13.2.1 is the last Xcode version that supports macOS Big Sur 11.

13.3 requires macOS Monterey 12 and while GitHub doesn't support macOS 12 yet, it will soon: https://github.com/actions/virtual-environments/issues/3649#issuecomment-1074058969.

ewanharris replied:
Similar to the Node 16 comment, I left everything as it to avoid introducing new things that could break while I was refactoring all this. I can update this to the latest available Xcode if desired

ewanharris commented on .github/workflows/build.yml:
```diff
@@ -4,111 +4,22 @@ on:
-      run: |
-        rm -f ~/.gradle/caches/modules-2/modules-2.lock
-        rm -f ~/.gradle/caches/modules-2/gc.properties
+      - uses: actions/checkout@v2
```
I don't see why not, that version just didn't exist when this PR was made

ewanharris replied:
Updated all actions/checkout uses to v3

ewanharris commented on .github/workflows/build.yml:
```diff
@@ -4,111 +4,22 @@ on:
+      - name: Android build
+        uses: ./.github/actions/build-android
+        with:
+          node-version: '12.x'
```
Maybe? I just stuck with what was there originally to ensure no other variables were at play when I was refactoring

ewanharris commented on releasing_the_sdk.md:
```diff
@@ -1,5 +1,7 @@
 # Releasing the SDK
 
+This document pertains to the Axway process for releasing the SDK, it will not be updated moving forward. Instead please refer to [this document](./maintainer-docs/releasing-the-sdk.md).
```
It was still relevant at the time of PR creation, will remove now that the repo has been handed over

cb1kenobi commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,165 @@
+    runs-on: ubuntu-latest
+    name: Release
+    environment: Release
+    needs: [validate, package]
```
30 minutes when doing a release is acceptable I would guess.

ewanharris commented on .github/actions/package/action.yml:
```diff
@@ -0,0 +1,106 @@
+      run: tar -xzvf ios-build.tar.gz
+      shell: bash
+
+    - name: Cache Native Modules
```
Well, an `npm run build` wont touch modules because it runs `scons build` which just builds for the requested platforms, it does not package the SDK which is where the module downloads are done (if you ran `npm run cleanbuild` or `npm run package` you would see this happen)

The modules are pulled down within [includePackagedModules](https://github.com/tidev/titanium_mobile/blob/master/build/lib/packager.js#L218), with how the scons scripts are written I guess yes something specific is done as that function is [called for each OS requested](https://github.com/tidev/titanium_mobile/blob/master/build/lib/builder.js#L220-L222)

ewanharris commented on .github/actions/build-android/action.yml:
```diff
@@ -0,0 +1,104 @@
+        java-version: ${{ inputs.java-version }}
+
+    - name: Install dependencies
+      run: npm ci
```
I don't think so, my understanding of the setup-node cache is that it's for the global modules directory not a projects node_modules. So that key tells us if setup-node restored the cached global node_modules directory not a local node_modules for a project

[ref](https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#caching-packages-dependencies)

ewanharris committed:
ci: move to v3 checkout action, use node 16

ewanharris commented on maintainer-docs/ci-setup.md:
```diff
@@ -0,0 +1,110 @@
+
+This is ran automatically on pull requests and on pushes to branches. There is no way to (and no need to) trigger this outside of those actions.
+
+This has no requirements
```
I intentionally wrote it here to make it explicit that there were no requirements given that other places declared their requirements. Missing period is just cos my grammar sucks

ewanharris committed:
docs: change org name in github links to tidev

ewanharris commented on .github/actions/package/action.yml:
```diff
@@ -0,0 +1,106 @@
+        java-version: ${{ inputs.java-version }}
+
+    - name: Install dependencies
+      if: steps.node-cache.outputs.cache-hit != 'true'
```
This was just copied over from the existing code I think, based on the reply to your previous comment I think I might actually remove it as I think it's not used correctly

ewanharris committed:
ci: remove cache hit check on dependency install

The setup-node cache is for global node_modules not local, so cache is not relevant

ewanharris commented on .github/actions/build-android/action.yml:
```diff
@@ -0,0 +1,104 @@
+
+    - name: Install dependencies
+      run: npm ci
+      shell: bash
```
Because these files are for [composite actions](https://github.blog/changelog/2021-08-25-github-actions-reduce-duplication-with-action-composition/) which are intended to be shared and reused (as we do here between the build and release workflows), you have linked to a workflow action file which is not the same thing. 

As the ADR I linked to states "The shell attribute is required for each run step because the action author does not know what the workflow author is using for the operating system so we need to explicitly prevent unknown behavior by making sure that each run step has an explicit shell set by the action author"

ewanharris committed:
docs: remove outdated release documentation

ewanharris committed:
ci: use Xcode 13.2.1

ewanharris committed:
ci(release): fix syntax issue

ewanharris commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,165 @@
+    steps:
+      - uses: actions/checkout@v2
+        with:
+            ref: ${{ github.event.inputs.branch }}
```
As evidenced by GitHub throwing errors on the file twice just now due to me screwing up yaml syntax, I hate yaml syntax 😄

cb1kenobi commented on .github/workflows/docs.yml:
```diff
@@ -12,12 +12,16 @@ jobs:
+        with:
+          node-version: '16.x'
+          cache: 'npm'
+      - run: npm ci
```
I'm still confused of when you need to say `shell: bash`. Should we add it here?

ewanharris replied:
No, it doesn't need to be added here as this is a workflow definition (i.e it is under `.github/workflows`)

cb1kenobi reviewed (approved):
I've tested both beta and GA releases and it works. I've created the Release environment.

It's a shame I can't see the artifacts until after the workflow has finished. Maybe there's a hack we can figure out later.

**APPROVED**!!!!
